internal/lang/funcs/encoding.go

```diff
@@ -227,6 +227,47 @@ var URLEncodeFunc = function.New(&function.Spec{
 	},
 })
 
+// URLDecodeFunc constructs a function that applies URL decoding to a given string.
+var URLDecodeFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "str",
+			Type: cty.String,
+		},
+	},
+	Type:         function.StaticReturnType(cty.String),
+	RefineResult: refineNotNull,
+	Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+		var reconstructedURL string
+		rawURL := args[0].AsString()
+		u, err := url.Parse(rawURL)
+		if err != nil {
+			return cty.UnknownVal(cty.String), fmt.Errorf("failed to parse URL '%s': %v", rawURL, err)
+		}
+		encodedQuery := u.Query()
+		if u.Scheme != "" {
+			reconstructedURL = u.Scheme + "://" + u.Host + u.Path
+		} else {
+			reconstructedURL = u.Host + u.Path
+		}
+
+		first := true
+		for key, values := range encodedQuery {
+			if len(values) > 0 {
+				if first {
+					reconstructedURL += "?"
+					first = false
+				} else {
+					reconstructedURL += "&"
+				}
+				reconstructedURL += key + "=" + values[0]
+			}
+		}
+
+		return cty.StringVal(reconstructedURL), nil
```

Comment:
Is there a reason you're not using url.QueryUnescape here? https://pkg.go.dev/net/url#QueryUnescape

Doing this manually feels a little hacky. I expected this to be much smaller, similar to the UrlEncode function.